aspnetcore/signalr/streaming.md

```diff
@@ -13,34 +13,52 @@ uid: signalr/streaming
+::: moniker-end
 
 ::: moniker range=">= aspnetcore-3.0"
 
-A hub method automatically becomes a streaming hub method when it returns a `ChannelReader<T>`, `IAsyncEnumerable<T>`, `Task<ChannelReader<T>>`, or `Task<IAsyncEnumerable<T>>`.
+ASP.NET Core SignalR supports streaming from client to server and from server to client. This is useful for scenarios where fragments of data will come in over time. When streaming, each fragment is sent to the client or server as soon as it becomes available, rather than waiting for all the data to become available.
 
 ::: moniker-end
 
-::: moniker range="< aspnetcore-3.0"
+::: moniker range="= aspnetcore-2.1"
 
-A hub method automatically becomes a streaming hub method when it returns a `ChannelReader<T>` or a `Task<ChannelReader<T>>`.
+[View or download sample code](https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming/sample.netcoreapp2.1) ([how to download](xref:index#how-to-download-a-sample))
+
+::: moniker-end
+
+::: moniker range="= aspnetcore-2.2"
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming/sample.netcoreapp2.2) ([how to download](xref:index#how-to-download-a-sample))
 
 ::: moniker-end
 
 ::: moniker range=">= aspnetcore-3.0"
 
-In ASP.NET Core 3.0 or later, streaming hub methods can return `IAsyncEnumerable<T>` in addition to `ChannelReader<T>`. The simplest way to return `IAsyncEnumerable<T>` is by making the hub method an async iterator method as the following sample demonstrates. Hub async iterator methods can accept a `CancellationToken` parameter that will be triggered when the client unsubscribes from the stream. Async iterator methods easily avoid problems common with Channels such as not returning the `ChannelReader` early enough or exiting the method without completing the `ChannelWriter`.
+[View or download sample code](https://github.com/aspnet/Docs/tree/live/aspnetcore/signalr/streaming/sample.netcoreapp3.0) ([how to download](xref:index#how-to-download-a-sample))
 
-[!INCLUDE[](~/includes/csharp-8-required.md)]
+::: moniker-end
+
+::: moniker range=">= aspnetcore-3.0"
+
+A hub method automatically becomes a streaming hub method when it returns a `ChannelReader<T>`, `IAsyncEnumerable<T>`, `Task<ChannelReader<T>>`, or `Task<IAsyncEnumerable<T>>`.
 
-[!code-csharp[Streaming hub async iterator method](streaming/sample/Hubs/AsyncEnumerableHub.cs?name=snippet_AsyncIterator)]
+::: moniker-end
+
+::: moniker range="< aspnetcore-3.0"
+
+A hub method automatically becomes a streaming hub method when it returns a `ChannelReader<T>` or a `Task<ChannelReader<T>>`.
 
 ::: moniker-end
 
+## Set up the hub for streaming
+
+### Server to client streaming
+
 The following sample shows the basics of streaming data to the client using Channels. Whenever an object is written to the `ChannelWriter` that object is immediately sent to the client. At the end, the `ChannelWriter` is completed to tell the client the stream is closed.
```

Comment:
We use ``<xref:System.Threading.Channels.ChannelReader`1>`` below so we should probably also reference ``<xref:System.Threading.Channels.ChannelWriter`1>`` here.